### PR TITLE
Hlt hcal movemodulestol1t

### PR DESCRIPTION
**Pull request by ssekmen**
Adapting two HCAL modules to use l1t instead of l1extra.

Created two new modules 
IsolatedPixelTrackCandidateProducer -> IsolatedPixelTrackCandidateL1TProducer
HLTPixelIsolTrackFilter -> HLTPixelIsolTrackL1TFilter
which replace l1extra jet objects used for taus with l1t tau objects.

This also required the addition of a new constructor to 
DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h
which uses l1t taus.

### CODE REVIEW

**Comment by Martin-Grunewald** [comment]
-1
please rebase to get rid of the other stuff

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
The DEFINE_FWK_MODULE calls for the new plugins are missing.
The entry in classes.h and class_def.xml for the new DataFormat class is missing
I doubt you could have tested the code properly without that...

File: DataFormats/HcalIsolatedTrack/interface/IsolatedL1TPixelTrackCandidate.h [hunk 1]
```diff
@@ -0,0 +1,152 @@
+#ifndef HcalIsolatedTrack_IsolatedPixelTrackCandidate_h
+#define HcalIsolatedTrack_IsolatedPixelTrackCandidate_h
+/** \class reco::IsolatedPixelTrackCandidate
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedL1TPixelTrackCandidate.h]: The "L1T" is missing in the above 3 lines...

Where is the *.cc implementation file? There is one for the old class src/IsolatedL1TPixelTrackCandidate.cc,
so I would expect one for the new one as well.

File: DataFormats/HcalIsolatedTrack/interface/IsolatedL1TPixelTrackCandidate.h [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+#include "DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidateFwd.h"
+
+#include <vector>
+#include <map>
+#include <utility>
+
+namespace reco {
+  
+  class IsolatedPixelTrackCandidate: public RecoCandidate {
```
Comment (slava77) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedL1TPixelTrackCandidate.h]: looks like replace step of  IsolatedPixelTrackCandidate -> IsolatedL1TPixelTrackCandidate failed in this file.
Do we really have to generate a new data format for this?

**Comment by slava77** [comment]
-1
- name conflicts (incomplete renaming in IsolatedL1TPixelTrackCandidate.h)
- consider staying with just IsolatedPixelTrackCandidate as an abstraction to the possible algorithms that create this data format. This way downstream code changes will be reduced.

**Comment by ssekmen** [comment]
@slava77 , is it a better idea to keep the old dataformat with the old constructor with l1extra, and add a new constructor that uses l1t:JetRef?
Thanks.

**Comment by slava77** [comment]
On 3/5/16 12:41 AM, Sezen Sekmen wrote:
> @slava77 <https://github.com/slava77> , is it a better idea to keep the
> old dataformat with the old constructor with l1extra, and add a new
> constructor that uses l1t:JetRef?

Yes.

File: DataFormats/HcalIsolatedTrack/interface/IsolatedL1TPixelTrackCandidate.h [hunk 3]
```diff
@@ -0,0 +1,152 @@
+#ifndef HcalIsolatedTrack_IsolatedPixelTrackCandidate_h
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedL1TPixelTrackCandidate.h]: If this file is no longer needed, then it must be removed.

File: Calibration/HcalIsolatedTrackReco/interface/IsolatedPixelTrackCandidateL1TProducer.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
+#ifndef Calibration_IsolatedPixelTrackCandidateL1TProducer_h
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/interface/IsolatedPixelTrackCandidateL1TProducer.h]: The new class must still be declared as a plugin.

**Comment by Martin-Grunewald** [comment]
@ssekmen 
Please make the required changes asap.
Thanks

**Comment by Martin-Grunewald** [comment]
please test

File: Calibration/HcalIsolatedTrackReco/interface/IsolatedPixelTrackCandidateL1TProducer.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+private:
+  struct seedAtEC {
+    seedAtEC(unsigned int i, bool f, double et, double fi) : index(i), ok(f), eta(et), phi(fi) { }
+    unsigned int index;
+    bool         ok;
+    double       eta, phi;
+  };
+
+  const edm::EDGetTokenT<trigger::TriggerFilterObjectWithRefs> tok_hlt_;
+  const edm::EDGetTokenT<l1t::JetBxCollection> tok_l1_;
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/interface/IsolatedPixelTrackCandidateL1TProducer.h]: This must become l1t::TauBxCollection (Taus now have a different type than just reusing the Jet type)

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+#include "Geometry/EcalAlgo/interface/EcalBarrelGeometry.h"
+#include "Geometry/EcalAlgo/interface/EcalEndcapGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloSubdetectorGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+
+
+IsolatedPixelTrackCandidateL1TProducer::IsolatedPixelTrackCandidateL1TProducer(const edm::ParameterSet& config) :
+  tok_hlt_(                     consumes<trigger::TriggerFilterObjectWithRefs>(config.getParameter<edm::InputTag>("L1GTSeedLabel")) ),
+  tok_l1_(                      consumes<l1t::JetBxCollection>(config.getParameter<edm::InputTag>("L1eTauJetsSource")) ),
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: l1t::TauBxCollection

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+  for (unsigned int iPix=0; iPix<toks_pix_.size(); iPix++) {
+    edm::Handle<reco::TrackCollection> iPixCol;
+    theEvent.getByToken(toks_pix_[iPix],iPixCol);
+    for (reco::TrackCollection::const_iterator pit=iPixCol->begin(); pit!=iPixCol->end(); pit++) {
+      pixelTrackRefs.push_back(reco::TrackRef(iPixCol,pit-iPixCol->begin()));
+    }
+  }
+
+  edm::Handle<l1t::JetBxCollection> l1eTauJets;
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Tau

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+  theEvent.getByToken(tok_vert_,pVert);
+
+  double ptTriggered  = -10;
+  double etaTriggered = -100;
+  double phiTriggered = -100;
+  
+  edm::Handle<trigger::TriggerFilterObjectWithRefs> l1trigobj;
+  theEvent.getByToken(tok_hlt_, l1trigobj);
+  
+  std::vector< edm::Ref<l1t::JetBxCollection> > l1tauobjref;
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Tau

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+  std::vector< edm::Ref<l1t::JetBxCollection> > l1jetobjref;
+  std::vector< edm::Ref<l1t::JetBxCollection> > l1forjetobjref;
+  
+  l1trigobj->getObjects(trigger::TriggerL1TauJet, l1tauobjref);
+  l1trigobj->getObjects(trigger::TriggerL1CenJet, l1jetobjref);
+  l1trigobj->getObjects(trigger::TriggerL1ForJet, l1forjetobjref);
+
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Cen and For jet no longer exists.
The enums have become TriggerTau and TriggerJet

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    //select tracks not matched to triggered L1 jet
+    double R=reco::deltaR(etaTriggered, phiTriggered, 
+			  pixelTrackRefs[iS]->eta(), pixelTrackRefs[iS]->phi());
+    if (R<tauUnbiasCone_) continue;
+
+    //check taujet matching
+    bool tmatch=false;
+    l1t::JetBxCollection::const_iterator selj;
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Tau

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+    for (l1t::JetBxCollection::const_iterator tj=l1eTauJets->begin(); tj!=l1eTauJets->end(); tj++) {
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Tau

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+    else       seedCooAtEC=GetEtaPhiAtEcal(pixelTrackRefs[iS]->eta(), pixelTrackRefs[iS]->phi(), pixelTrackRefs[iS]->pt(), pixelTrackRefs[iS]->charge(), 0);
+    seedAtEC seed(iS,(tmatch||vtxMatch),seedCooAtEC.first,seedCooAtEC.second);
+    VecSeedsatEC.push_back(seed);
+  }
+
+  for (unsigned int i=0; i<VecSeedsatEC.size(); i++) {
+    unsigned int iSeed = VecSeedsatEC[i].index;
+    if (!VecSeedsatEC[i].ok) continue;
+    if(pixelTrackRefs[iSeed]->p()<minPTrackValue_) continue; 
+    l1t::JetBxCollection::const_iterator selj;
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Tau

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+    for (l1t::JetBxCollection::const_iterator tj=l1eTauJets->begin(); tj!=l1eTauJets->end(); tj++)  {
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Tau

File: Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+      //cut ot dXY:
+      if (found&&std::abs(pixelTrackRefs[iSurr]->dxy(vitSel2->position()))>vtxCutIsol_) continue;
+      //calculate distance at ECAL surface and update isolation: 
+      if (getDistInCM(VecSeedsatEC[i].eta, VecSeedsatEC[i].phi, VecSeedsatEC[j].eta, VecSeedsatEC[j].phi)<pixelIsolationConeSizeAtEC_) {
+	sumP+=pixelTrackRefs[iSurr]->p();
+	if(pixelTrackRefs[iSurr]->p()>maxP) maxP=pixelTrackRefs[iSurr]->p();
+      }      
+    }
+    if (maxP<maxPForIsolationValue_) {
+      reco::IsolatedPixelTrackCandidate newCandidate(pixelTrackRefs[iSeed], l1t::JetRef(l1eTauJets,selj-l1eTauJets->begin()), maxP, sumP);
```
Comment (Martin-Grunewald) [inline Calibration/HcalIsolatedTrackReco/src/IsolatedPixelTrackCandidateL1TProducer.cc]: Tau

File: DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h [hunk 15]
```diff
@@ -58,6 +59,18 @@ namespace reco {
       phiEcal_=0;
       etaPhiEcal_=false;
     }
+    // constructor from a track using l1t
+  IsolatedPixelTrackCandidate(const reco::TrackRef & tr, const l1t::JetRef & tauRef, double max, double sum):
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h]: I suppose Tau here as well (or do you really want [also] Jets?)

File: DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h [hunk 16]
```diff
@@ -69,6 +82,16 @@ namespace reco {
       phiEcal_=0;
       etaPhiEcal_=false;
     }
+    ///constructor from tau jet using l1t
+    IsolatedPixelTrackCandidate(const l1t::JetRef & tauRef, double enIn, double enOut, int nhitIn, int nhitOut):
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h]: TauRef

File: DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h [hunk 17]
```diff
@@ -93,7 +116,11 @@ namespace reco {
     /// get reference to L1 tau jet
     virtual l1extra::L1JetParticleRef l1tau() const;
     void    setL1TauJet( const l1extra::L1JetParticleRef & tauRef ) { l1tauJet_ = tauRef; }
-          
+
+    /// get reference to L1 tau jet from lt1
+    virtual l1t::JetRef l1ttau() const;
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h]: Tau

File: DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h [hunk 18]
```diff
@@ -129,6 +156,8 @@ namespace reco {
     reco::TrackRef track_;
     /// reference to a L1 tau jet
     l1extra::L1JetParticleRef l1tauJet_;
+    /// reference to a S2 L1 tau jet
+    l1t::JetRef l1ttauJet_;
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h]: l1t::TauRef

**Comment by Martin-Grunewald** [comment]
Change Jet to Tau in MOST places.
Why on earth did you not read https://github.com/cms-l1t-offline/cmssw/issues/156

**Comment by Martin-Grunewald** [comment]
-1

File: DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h [hunk 19]
```diff
@@ -93,7 +116,11 @@ namespace reco {
     /// get reference to L1 tau jet
     virtual l1extra::L1JetParticleRef l1tau() const;
     void    setL1TauJet( const l1extra::L1JetParticleRef & tauRef ) { l1tauJet_ = tauRef; }
-          
+
+    /// get reference to L1 tau jet from lt1
+    virtual l1t::JetRef l1ttau() const;
+    void    setL1TTauJet( const l1t::JetRef & tauRef ) { l1ttauJet_ = tauRef; }
+
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h]: TauRef

File: DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h [hunk 20]
```diff
@@ -129,6 +156,8 @@ namespace reco {
     reco::TrackRef track_;
     /// reference to a L1 tau jet
     l1extra::L1JetParticleRef l1tauJet_;
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/interface/IsolatedPixelTrackCandidate.h]: L1TauParticleRef

File: DataFormats/HcalIsolatedTrack/src/IsolatedPixelTrackCandidate.cc [hunk 21]
```diff
@@ -29,6 +29,10 @@ l1extra::L1JetParticleRef IsolatedPixelTrackCandidate::l1tau() const {
   return l1tauJet_;
 }
 
+l1t::JetRef IsolatedPixelTrackCandidate::l1ttau() const {
+  return l1ttauJet_;
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/src/IsolatedPixelTrackCandidate.cc]: TauRef

File: HLTrigger/special/src/HLTPixelIsolTrackL1TFilter.cc [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+  edm::Handle<trigger::TriggerFilterObjectWithRefs> l1trigobj;
+  iEvent.getByToken(hltGTseedToken_, l1trigobj);
+
+//  std::vector< edm::Ref<l1t::L1JetParticleCollection> > l1tauobjref;
+//  std::vector< edm::Ref<l1t::L1JetParticleCollection> > l1jetobjref;
+//  std::vector< edm::Ref<l1t::L1JetParticleCollection> > l1forjetobjref;
+  std::vector< edm::Ref<l1t::JetBxCollection> > l1tauobjref;
+  std::vector< edm::Ref<l1t::JetBxCollection> > l1jetobjref;
+  std::vector< edm::Ref<l1t::JetBxCollection> > l1forjetobjref;
+
```
Comment (Martin-Grunewald) [inline HLTrigger/special/src/HLTPixelIsolTrackL1TFilter.cc]: As above - there are only Jet and Tau types and collections, each with its own C++type

File: HLTrigger/special/src/HLTPixelIsolTrackL1TFilter.cc [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+  l1trigobj->getObjects(trigger::TriggerL1TauJet, l1tauobjref);
+  l1trigobj->getObjects(trigger::TriggerL1CenJet, l1jetobjref);
+  l1trigobj->getObjects(trigger::TriggerL1ForJet, l1forjetobjref);
+
```
Comment (Martin-Grunewald) [inline HLTrigger/special/src/HLTPixelIsolTrackL1TFilter.cc]: See above

File: HLTrigger/special/src/HLTPixelIsolTrackL1TFilter.cc [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+  for (unsigned int p=0; p<l1tauobjref.size(); p++)
+    if (l1tauobjref[p]->pt() > ptTriggered)
+      ptTriggered = l1tauobjref[p]->pt();
+  for (unsigned int p=0; p<l1jetobjref.size(); p++)
+    if (l1jetobjref[p]->pt() > ptTriggered)
+      ptTriggered = l1jetobjref[p]->pt();
+  for (unsigned int p=0; p<l1forjetobjref.size(); p++)
+    if (l1forjetobjref[p]->pt() > ptTriggered)
+      ptTriggered = l1forjetobjref[p]->pt();
+
```
Comment (Martin-Grunewald) [inline HLTrigger/special/src/HLTPixelIsolTrackL1TFilter.cc]: See above

**Comment by ssekmen** [comment]
Thanks @Martin-Grunewald .  I did read cms-l1t-offline#156, but did not realize that a change of Jet->Tau was compulsory, which was mentioned in the later conversation.  
I'm making the recommended changes now.

File: DataFormats/HcalIsolatedTrack/src/classes_def.xml [hunk 25]
```diff
@@ -2,7 +2,8 @@
 
   <class name="edm::Wrapper<std::map <int,std::pair<double,double> > >"/> 
 
-  <class name="reco::IsolatedPixelTrackCandidate" ClassVersion="13">
+  <class name="reco::IsolatedPixelTrackCandidate" ClassVersion="14">
+   <version ClassVersion="14" checksum="2494226563"/>
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/src/classes_def.xml]: When making the new DataFormat, re-use classversion 14 but with the new checksum, as 14 was not yet used anywhere.

File: DataFormats/HcalIsolatedTrack/src/classes_def.xml [hunk 26]
```diff
@@ -2,7 +2,9 @@
 
   <class name="edm::Wrapper<std::map <int,std::pair<double,double> > >"/> 
 
-  <class name="reco::IsolatedPixelTrackCandidate" ClassVersion="13">
+  <class name="reco::IsolatedPixelTrackCandidate" ClassVersion="15">
+   <version ClassVersion="15" checksum="4036611259"/>
+   <version ClassVersion="14" checksum="2494226563"/>
    <version ClassVersion="13" checksum="2098110013"/>
```
Comment (Martin-Grunewald) [inline DataFormats/HcalIsolatedTrack/src/classes_def.xml]: Remove the line with 14, and change 15 to 14.

**Comment by Martin-Grunewald** [comment]
Please fix DataFormats/HcalIsolatedTrack/src/classes_def.xml

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/13596	98358ef
- code changes in the reco packages are to add construction of IsolatedPixelTrackCandidate.h with L1T tau
- jenkins tests pass and comparisons with baseline show no differences

@ssekmen please update the PR description (the first block of text starting with "Moving two HCAL modules") to appropriately describe changes in  DataFormats/HcalIsolatedTrack

**Comment by Martin-Grunewald** [comment]
@diguida, @cerminar @franzoni, @deguio
can you please sign this PR for AlCa?
Thanks!

**Comment by ssekmen** [comment]
@slava77 , thanks, I updated the PR description.

**Comment by mmusich** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**